File: ios/Sources/MoQKit/Publish/source/CaptureLifecycle.swift
import Foundation

/// Serial control executor shared by capture and publication. Frame callbacks never wait on it.
enum PublishControl {
    private static let key = DispatchSpecificKey<Bool>()
    static let queue: DispatchQueue = {
        let queue = DispatchQueue(label: "com.swmansion.MoQKit.publish-control")
        queue.setSpecific(key: key, value: true)
        return queue
    }()

    static func sync<T>(_ action: () throws -> T) rethrows -> T {
        if DispatchQueue.getSpecific(key: key) == true { return try action() }
        return try queue.sync(execute: action)
    }

    static func run<T>(_ action: @escaping () throws -> T) async throws -> T {
        try await withCheckedThrowingContinuation { continuation in
            queue.async { continuation.resume(with: Result { try action() }) }
        }
    }

    static func finish(_ action: @escaping () -> Void) async {
        await withCheckedContinuation { continuation in
            queue.async { action(); continuation.resume() }
        }
    }
}

/// Replays availability and delivers every transition on PublishControl.
final class CaptureLifecycle {
    struct Snapshot {
        var running = false
        var closed = false
        var generation: UInt64 = 0
    }
    private(set) var snapshot = Snapshot()
    private var observers: [UUID: (Snapshot) -> Void] = [:]
    private var owner: UUID?

    func reserve(_ id: UUID) throws {
        guard !snapshot.closed else { throw SessionError.alreadyClosed }
        guard owner == nil else {
            throw SessionError.invalidConfiguration("Capture already has a publication attachment")
        }
        owner = id
    }

    func release(_ id: UUID) {
        if owner == id { owner = nil }
    }

    func setRunning(_ value: Bool) {
        guard !snapshot.closed, snapshot.running != value else { return }
        snapshot.running = value
        if value { snapshot.generation &+= 1 }
        notify()
    }

    func close() {
        guard !snapshot.closed else { return }
        snapshot.running = false
        snapshot.closed = true
        notify()
        observers.removeAll()
        owner = nil
    }

    private func notify() {
        for observer in Array(observers.values) { observer(snapshot) }
    }

    func observe(_ observer: @escaping (Snapshot) -> Void) -> () -> Void {
        let id = UUID()
        observers[id] = observer
        observer(snapshot)
        return { [weak self] in self?.observers[id] = nil }
    }
}

/// All resource transitions for one media publication pass through reconcile().
final class CaptureTrackBinding {
    private var closed = false
    private var enabled: Bool
    private var source = CaptureLifecycle.Snapshot(running: true)
    private var generation: UInt64?
    private var stopEncoding: (() -> Void)?
    private var cancelObservation: (() -> Void)?
    private let startEncoding: () throws -> () -> Void
    private let onState: (PublishedTrackState) -> Void
    private let onError: (Error) -> Void
    private let onClosed: () -> Void
    private(set) var failure: Error?

    init(enabled: Bool, start: @escaping () throws -> () -> Void,
         onState: @escaping (PublishedTrackState) -> Void,
         onError: @escaping (Error) -> Void, onClosed: @escaping () -> Void) {
        self.enabled = enabled
        startEncoding = start
        self.onState = onState
        self.onError = onError
        self.onClosed = onClosed
    }

    func attach(to lifecycle: CaptureLifecycle?) throws {
        if let lifecycle {
            let cancel = lifecycle.observe { [weak self] snapshot in
                guard let self, !self.closed else { return }
                if snapshot.generation != self.source.generation { self.failure = nil }
                self.source = snapshot
                self.reconcile()
            }
            if closed { cancel() } else { cancelObservation = cancel }
        } else { reconcile() }
        if let failure { throw failure }
    }

    func setEnabled(_ value: Bool) throws {
        guard !closed else { throw SessionError.alreadyClosed }
        enabled = value
        failure = nil
        reconcile()
        if let failure { throw failure }
    }

    func fail(_ error: Error) {
        guard !closed else { return }
        stopOutput()
        failure = error
        onState(.failed(error.localizedDescription))
        onError(error)
    }

    private func reconcile() {
        guard !closed else { return }
        if source.closed { stop(); onClosed(); return }
        if !enabled || !source.running {
            stopOutput()
            onState(enabled ? .idle : .disabled)
            return
        }
        if generation != source.generation { stopOutput() }
        guard stopEncoding == nil, failure == nil else { return }
        do {
            onState(.starting)
            stopEncoding = try startEncoding()
            generation = source.generation
        } catch { fail(error) }
    }

    private func stopOutput() {
        let stop = stopEncoding
        stopEncoding = nil
        generation = nil
        stop?()
    }

    func stop() {
        guard !closed else { return }
        closed = true
        stopOutput()
        cancelObservation?()
        cancelObservation = nil
    }
}

/// Fences encoder input against teardown without holding a lock while draining the codec.
final class CaptureFrameGate: @unchecked Sendable {
    private let lock = NSLock()
    private var closed = false

    func send(_ encode: () -> Void) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        guard !closed else { return false }
        encode()
        return true
    }

    func close(_ stop: () -> Void) {
        lock.lock()
        closed = true
        lock.unlock()
        stop()
    }
}

File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/publish/source/CaptureLifecycle.kt
package com.swmansion.moqkit.publish.source

import com.swmansion.moqkit.publish.PublishedTrackState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext

/** One ordered control lane. Codec/frame callbacks enqueue notifications without waiting. */
internal object PublishControl {
    private val mutex = Mutex()
    private val notifications = CoroutineScope(SupervisorJob() + Dispatchers.Default.limitedParallelism(1))

    suspend fun <T> run(action: suspend () -> T): T =
        withContext(NonCancellable + Dispatchers.Default) { mutex.withLock { action() } }

    fun post(action: suspend () -> Unit) {
        notifications.launch { mutex.withLock { action() } }
    }
}

/** Replays current availability; changes and observers are confined to PublishControl. */
internal class CaptureLifecycle {
    data class Snapshot(val running: Boolean = false, val closed: Boolean = false, val generation: Long = 0)
    @Volatile var snapshot = Snapshot()
        private set
    private var owner: Any? = null
    private val observers = mutableMapOf<Any, (Snapshot) -> Unit>()

    @Synchronized fun reserve(id: Any) {
        check(!snapshot.closed) { "Capture is closed" }
        check(owner == null) { "Capture already has a publication attachment" }
        owner = id
    }

    @Synchronized fun release(id: Any) {
        if (owner === id) owner = null
    }

    fun setRunning(value: Boolean) {
        if (snapshot.closed || snapshot.running == value) return
        snapshot = snapshot.copy(running = value, generation = snapshot.generation + if (value) 1 else 0)
        observers.values.toList().forEach { it(snapshot) }
    }

    fun close() {
        if (snapshot.closed) return
        synchronized(this) {
            snapshot = snapshot.copy(running = false, closed = true)
            owner = null
        }
        observers.values.toList().forEach { it(snapshot) }
        observers.clear()
    }

    fun observe(observer: (Snapshot) -> Unit): () -> Unit {
        val id = Any()
        observers[id] = observer
        observer(snapshot)
        return { observers.remove(id); Unit }
    }
}

/** The sole owner of activation/deactivation for a logical media publication. */
internal class CaptureTrackBinding(
    private var enabled: Boolean,
    private val start: () -> (() -> Unit),
    private val onState: (PublishedTrackState) -> Unit,
    private val onError: (Exception) -> Unit,
    private val onClosed: () -> Unit,
) {
    private var closed = false
    private var source = CaptureLifecycle.Snapshot(running = true)
    private var generation: Long? = null
    private var stopEncoding: (() -> Unit)? = null
    private var cancelObservation: (() -> Unit)? = null
    var failure: Exception? = null
        private set

    fun attach(lifecycle: CaptureLifecycle?) {
        if (lifecycle == null) reconcile()
        else {
            val cancel = lifecycle.observe { snapshot ->
                if (!closed) {
                    if (snapshot.generation != source.generation) failure = null
                    source = snapshot
                    reconcile()
                }
            }
            if (closed) cancel() else cancelObservation = cancel
        }
        failure?.let { throw it }
    }

    fun setEnabled(value: Boolean) {
        check(!closed) { "Track is stopped" }
        enabled = value
        failure = null
        reconcile()
        failure?.let { throw it }
    }

    fun fail(error: Exception) {
        if (closed) return
        stopOutput()
        failure = error
        onState(PublishedTrackState.Failed(error.message ?: "Media publication failed"))
        onError(error)
    }

    private fun reconcile() {
        if (closed) return
        if (source.closed) { stop(); onClosed(); return }
        if (!enabled || !source.running) {
            stopOutput()
            onState(if (enabled) PublishedTrackState.Idle else PublishedTrackState.Disabled)
            return
        }
        if (generation != source.generation) stopOutput()
        if (stopEncoding != null || failure != null) return
        try {
            onState(PublishedTrackState.Starting)
            stopEncoding = start()
            generation = source.generation
        } catch (e: Exception) { fail(e) }
    }

    private fun stopOutput() {
        val stop = stopEncoding
        stopEncoding = null
        generation = null
        try { stop?.invoke() } catch (e: Exception) { onError(e) }
    }

    fun stop() {
        if (closed) return
        closed = true
        stopOutput()
        cancelObservation?.invoke()
        cancelObservation = null
    }
}
